File: MainApp/src/main/java/com/sleticalboy/dailywork/accounts/auth/AuthenticatorActivity.java
package com.sleticalboy.dailywork.accounts.auth;

import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.AccountManager;
import android.content.ContentResolver;
import android.content.Intent;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.sleticalboy.dailywork.R;
import com.sleticalboy.dailywork.accounts.Constants;
import com.sleticalboy.dailywork.base.BaseActivity;

import org.jetbrains.annotations.Nullable;

public class AuthenticatorActivity extends BaseActivity implements Constants {

    private static final String TAG = "AuthenticatorActivity";

    private AccountAuthenticatorResponse mAccountAuthenticatorResponse;
    private Bundle mResultBundle;
    private String mUsername, mPassword;
    private AccountManager mAccountManager;
    private boolean mRequestNewAccount = false;
    private ProgressBar mLoading;

    @Override
    protected void prepareWork(@Nullable final Bundle savedInstanceState) {
        mAccountManager = AccountManager.get(getApplication());
        mUsername = getIntent().getStringExtra(AccountManager.KEY_ACCOUNT_NAME);
        mRequestNewAccount = mUsername == null;
        mAccountAuthenticatorResponse = getIntent().getParcelableExtra(
                AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE);
        if (mAccountAuthenticatorResponse != null) {
            mAccountAuthenticatorResponse.onRequestContinued();
        }
    }

    @Override
    public void initView() {
        mLoading = findViewById(R.id.loading);
        final Button loginBtn = findViewById(R.id.login);
        final EditText usernameEt = findViewById(R.id.username);
        final EditText passwordEt = findViewById(R.id.password);
        final TextWatcher watcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start,
                                          final int count, final int after) {
            }

            @Override
            public void onTextChanged(final CharSequence s, final int start,
                                      final int before, final int count) {
            }

            @Override
            public void afterTextChanged(final Editable s) {
                loginBtn.setEnabled(!TextUtils.isEmpty(usernameEt.getText())
                        && !TextUtils.isEmpty(passwordEt.getText()));
            }
        };
        usernameEt.addTextChangedListener(watcher);
        passwordEt.addTextChangedListener(watcher);
        loginBtn.setOnClickListener(v -> {
            mLoading.setVisibility(View.VISIBLE);
            mUsername = usernameEt.getText().toString().trim();
            mPassword = passwordEt.getText().toString().trim();
            handleLogin(v);
        });
    }

    private void handleLogin(final View v) {
        if (mUsername.startsWith(ACCOUNT_PREFIX) && mPassword.startsWith(ACCOUNT_PASSWORD)) {
            v.postDelayed(() -> finishLogin(ACCOUNT_AUTH_TOKEN), 300L);
        } else {
            finishLogin(null);
        }
    }

    private void finishLogin(String authToken) {
        Log.d(TAG, "finishLogin() called with: authToken = [" + authToken + "]");
        if (authToken == null || authToken.length() == 0) {
            mLoading.setVisibility(View.GONE);
            Toast.makeText(this, R.string.login_failed, Toast.LENGTH_SHORT).show();
            return;
        }
        final Account account = new Account(mUsername, ACCOUNT_TYPE);
        if (mRequestNewAccount) {
            mAccountManager.addAccountExplicitly(account, mPassword, null);
            // Set contacts sync for this account.
            ContentResolver.setSyncAutomatically(account, ContactsContract.AUTHORITY, true);
        } else {
            mAccountManager.setPassword(account, mPassword);
        }
        final Intent intent = new Intent();
        intent.putExtra(AccountManager.KEY_ACCOUNT_NAME, mUsername);
        intent.putExtra(AccountManager.KEY_ACCOUNT_TYPE, ACCOUNT_TYPE);
        mResultBundle = intent.getExtras();
        setResult(RESULT_OK, intent);
        mLoading.setVisibility(View.GONE);
        finish();
    }

    @Override
    protected int layoutResId() {
        return R.layout.activity_login;
    }

    @Override
    public void finish() {
        if (mAccountAuthenticatorResponse != null) {
            // send the result bundle back if set, otherwise send an error.
            if (mResultBundle != null) {
                mAccountAuthenticatorResponse.onResult(mResultBundle);
            } else {
                mAccountAuthenticatorResponse.onError(AccountManager.ERROR_CODE_CANCELED,
                        "canceled");
            }
            mAccountAuthenticatorResponse = null;
        }
        super.finish();
    }
}

File: MainApp/src/main/java/com/sleticalboy/learning/accounts/auth/AuthenticatorActivity.kt
package com.sleticalboy.learning.accounts.auth

import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.content.ContentResolver
import android.content.Intent
import android.os.Bundle
import android.provider.ContactsContract
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.Toast
import com.sleticalboy.learning.R
import com.sleticalboy.learning.accounts.Constants
import com.sleticalboy.learning.base.BaseActivity

class AuthenticatorActivity : BaseActivity(), Constants {

    private var mAccountAuthenticatorResponse: AccountAuthenticatorResponse? = null
    private var mResultBundle: Bundle? = null
    private var mUsername: String? = null
    private var mPassword: String? = null
    private var mAccountManager: AccountManager? = null
    private var mRequestNewAccount = false
    private var mLoading: ProgressBar? = null

    override fun prepareWork(savedInstanceState: Bundle?) {
        mAccountManager = AccountManager.get(application)
        mUsername = intent.getStringExtra(AccountManager.KEY_ACCOUNT_NAME)
        mRequestNewAccount = mUsername == null
        mAccountAuthenticatorResponse = intent.getParcelableExtra(
                AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE)
        if (mAccountAuthenticatorResponse != null) {
            mAccountAuthenticatorResponse!!.onRequestContinued()
        }
    }

    public override fun initView() {
        mLoading = findViewById(R.id.loading)
        val loginBtn = findViewById<Button>(R.id.login)
        val usernameEt = findViewById<EditText>(R.id.username)
        val passwordEt = findViewById<EditText>(R.id.password)
        val watcher: TextWatcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int,
                                           count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence, start: Int,
                                       before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable) {
                loginBtn.isEnabled = (!TextUtils.isEmpty(usernameEt.text)
                        && !TextUtils.isEmpty(passwordEt.text))
            }
        }
        usernameEt.addTextChangedListener(watcher)
        passwordEt.addTextChangedListener(watcher)
        loginBtn.setOnClickListener { v: View ->
            mLoading?.visibility = View.VISIBLE
            mUsername = usernameEt.text.toString().trim { it <= ' ' }
            mPassword = passwordEt.text.toString().trim { it <= ' ' }
            handleLogin(v)
        }
    }

    private fun handleLogin(v: View) {
        if (mUsername!!.startsWith(Constants.ACCOUNT_PREFIX) && mPassword!!.startsWith(Constants.ACCOUNT_PASSWORD)) {
            v.postDelayed({ finishLogin(Constants.ACCOUNT_AUTH_TOKEN) }, 300L)
        } else {
            finishLogin(null)
        }
    }

    private fun finishLogin(authToken: String?) {
        Log.d(TAG, "finishLogin() called with: authToken = [$authToken]")
        if (authToken == null || authToken.isEmpty()) {
            mLoading!!.visibility = View.GONE
            Toast.makeText(this, R.string.login_failed, Toast.LENGTH_SHORT).show()
            return
        }
        val account = Account(mUsername, Constants.ACCOUNT_TYPE)
        if (mRequestNewAccount) {
            mAccountManager!!.addAccountExplicitly(account, mPassword, null)
            // Set contacts sync for this account.
            ContentResolver.setSyncAutomatically(account, ContactsContract.AUTHORITY, true)
        } else {
            mAccountManager!!.setPassword(account, mPassword)
        }
        val intent = Intent()
        intent.putExtra(AccountManager.KEY_ACCOUNT_NAME, mUsername)
        intent.putExtra(AccountManager.KEY_ACCOUNT_TYPE, Constants.ACCOUNT_TYPE)
        mResultBundle = intent.extras
        setResult(RESULT_OK, intent)
        mLoading!!.visibility = View.GONE
        finish()
    }

    override fun layoutResId(): Int {
        return R.layout.activity_login
    }

    override fun finish() {
        if (mAccountAuthenticatorResponse != null) {
            // send the result bundle back if set, otherwise send an error.
            if (mResultBundle != null) {
                mAccountAuthenticatorResponse!!.onResult(mResultBundle)
            } else {
                mAccountAuthenticatorResponse!!.onError(AccountManager.ERROR_CODE_CANCELED,
                        "canceled")
            }
            mAccountAuthenticatorResponse = null
        }
        super.finish()
    }

    companion object {
        private const val TAG = "AuthenticatorActivity"
    }
}